lib/pos_blockchain.go

```diff
@@ -408,20 +406,50 @@ func (bc *Blockchain) addBlockToBlockIndex(desoBlock *MsgDeSoBlock) error {
-// this block has a higher QC than our current tip and extends from either the committed
-// tip OR any uncommitted safe block in our block index.
-func (bc *Blockchain) shouldReorg(desoBlock *MsgDeSoBlock) bool {
-	return false
+// this block is proposed in a view greater than or equal to the currentView. Other
+// functions have validated that this block is not extending from a committed block
+// that is not the latest committed block, so there is no need to validate that here.
+func (bc *Blockchain) shouldReorg(desoBlock *MsgDeSoBlock, currentView uint64) bool {
+	chainTip := bc.GetBestChainTip()
+	// If this block extends from the chain tip, there's no need to reorg.
+	if chainTip.Hash.IsEqual(desoBlock.Header.PrevBlockHash) {
+		return false
+	}
+	// If the block is proposed in a view less than the current view, there's no need to reorg.
+	return desoBlock.Header.ProposedInView >= currentView
 }
 
-// handleReorg handles a reorg to the block provided. It does not check whether or not we should
+// tryReorgToNewTip handles a reorg to the block provided. It does not check whether or not we should
 // perform a reorg, so this should be called after shouldReorg. It will do the following:
-// 1. Update the bestChain and bestChainMap by removing blocks that are not uncommitted ancestor of this block.
+// 1. Update the bestChain and bestChainMap by removing blocks that are uncommitted and are not ancestors of this block.
 // 2. Update the bestChain and bestChainMap by adding blocks that are uncommitted ancestors of this block.
 // Note: addBlockToBestChain will be called after this to handle adding THIS block to the best chain.
-func (bc *Blockchain) handleReorg(desoBlock *MsgDeSoBlock) error {
-	// TODO: Implement me.
-	return errors.New("IMPLEMENT ME")
+func (bc *Blockchain) tryReorgToNewTip(desoBlock *MsgDeSoBlock, currentView uint64, ancestorsToCommittedTip []*BlockNode) (_hasReorg bool, _err error) {
+	// Check if we should perform a reorg here
+	if !bc.shouldReorg(desoBlock, currentView) {
+		return false, nil
+	}
+	// For simplicity, we remove all uncommitted blocks and then re-add them.
+	highestCommittedBlock, idx := bc.getHighestCommittedBlock()
+	if highestCommittedBlock == nil {
+		// This is an edge case we'll never hit in practice since all the PoW blocks
+		// are committed.
+		return false, errors.New("tryReorgToNewTip: No committed blocks found")
+	}
+
+	// Remove all uncommitted blocks. These are all blocks that come after the highestCommittedBlock
+	// in the best chain.
+	// Delete all blocks from bc.bestChainMap that come after the highest committed block.
+	for ii := idx + 1; ii < len(bc.bestChain); ii++ {
+		delete(bc.bestChainMap, *bc.bestChain[ii].Hash)
+	}
+	// Shorten best chain back to committed tip.
+	bc.bestChain = bc.bestChain[:idx+1]
+
+	for ii := len(ancestorsToCommittedTip) - 1; ii >= 0; ii-- {
+		bc.addBlockToBestChain(ancestorsToCommittedTip[ii])
+	}
+	return true, nil
```

Comment:
Do you want to add the incoming block to the `uncommittedBlocks` map as well? We may as well perform all changes to that map from one place. It seems natural to perform them here since the best chain is also being updated